src/tasks/WasmAppBuilder/EmitWasmBundleFiles.cs

```diff
@@ -0,0 +1,270 @@
+            if(string.IsNullOrEmpty(registeredName))
+            {
+                registeredName = Path.GetFileName(file.ItemSpec);
+            }
+            return registeredName;
+        }).ToList();
+
+        var files = filesToBundleByRegisteredName.Select(group => {
+            var registeredFile = group.First();
+            var outputFile = registeredFile.GetMetadata("DestinationFile");
+            var registeredName = group.Key;
+            var symbolName = ToSafeSymbolName(outputFile);
+            return (registeredName, symbolName);
+        }).ToList();
+
+        Log.LogMessage(MessageImportance.High, "Bundling {2} objects into mono_wasm_register_{0}_bundle as {1}", BundleName, BundleFile, files.Count);
+        emitter(BundleFile, (inputStream) =>
+        {
+            using var outputUtf8Writer = new StreamWriter(inputStream, Utf8NoBom);
+            GenerateRegisterBundledObjects($"mono_wasm_register_{BundleName}_bundle", RegistrationCallbackFunctionName, files, outputUtf8Writer);
+        });
+
+        return !Log.HasLoggedErrors;
+    }
+
+    public void Cancel()
+    {
+        BuildTaskCancelled.Cancel();
+    }
+
+    #region Helpers
+
+    private static readonly Encoding Utf8NoBom = new UTF8Encoding(false);
+    private static readonly byte[] HexToUtf8Lookup = InitLookupTable();
+    private static readonly byte[] NewLineAndIndentation = new[] { (byte)0x0a, (byte)0x20, (byte)0x20 };
+
+    private static byte[] InitLookupTable()
+    {
+        // Every 6 bytes in this array represents the output for a different input byte value.
+        // For example, the input byte 0x1a (26 decimal) corresponds to bytes 156-161 (26*6=156),
+        // whose values will be ['0', 'x', '1', 'a', ',', ' '], which is the UTF-8 representation
+        // for "0x1a, ". This is just a faster alternative to calling .ToString("x2") on every
+        // byte of the input file and then pushing that string through UTF8Encoding.
+        var lookup = new byte[256 * 6];
+        for (int i = 0; i < 256; i++)
+        {
+            string byteAsHex = i.ToString("x2");
+            char highOrderChar = BitConverter.IsLittleEndian ? byteAsHex[0] : byteAsHex[1];
+            char lowOrderChar = BitConverter.IsLittleEndian ? byteAsHex[1] : byteAsHex[0];
+            lookup[i * 6 + 0] = (byte)'0';
+            lookup[i * 6 + 1] = (byte)'x';
+            lookup[i * 6 + 2] = (byte)highOrderChar;
+            lookup[i * 6 + 3] = (byte)lowOrderChar;
+            lookup[i * 6 + 4] = (byte)',';
+            lookup[i * 6 + 5] = (byte)' ';
+        }
+
+        return lookup;
+    }
+
+    public void WriteSource(string destinationFile, Action<Stream> inputProvider)
+    {
+        using (var fileStream = File.Create(destinationFile))
+        {
+            inputProvider(fileStream);
+        }
+    }
+
+    public void Compile(string destinationFile, Action<Stream> inputProvider)
```

Comment:
This should return a bool, so we can stop the parallel loop. See the original task for that.